### PR TITLE
Implementation of signature based methods. 

### PR DESCRIPTION
**Pull request by jambo6**
This PR implements two classes for time series analysis with signatures:

- `GeneralisedSignatureMethod` Transformation class for generating signatures from time-series data.
- `SignatureClassifier` Wraps the above method with a tabular ML classifier.

The structure of the `GeneralisedSignatureMethod` follows closely from [this work in this paper](https://arxiv.org/pdf/2006.00873.pdf).

This adds esig as a soft dependency.

### CODE REVIEW

**Comment by TonyBagnall** [comment]
hi, thanks for updating this PR, its failing some of the automated tests. I'll discuss this with Markus and see if we can suggest solutions, I'd really like this in the toolkit

**Comment by jambo6** [comment]
Sorry it's taking ages, there have been a host of issues with each different signature package I tried. 

I should be one esig update away from tests passing though. The only remaining issue is some import fault in esig rather than an installation problem. The developers are aware and it should be sorted in the next update. 

Give me two weeks, then if tests are still not passing we should discuss.

**Comment by TonyBagnall** [comment]
no problem, I feel you pain :) And no rush, but if you need assistance give us a shout.

**Comment by jambo6** [comment]
At long last, everything seems to be working. However, as can be seen, it's not finishing the tests for macos. Do you know whats going wrong here?

**Comment by mloning** [comment]
Hi @jambo6, GitHub now requires approval from first-time contributors to run CI checks, just approved it now.

**Comment by fkiraly** [comment]
Quick question: is `SignatureClassifier` not redundant with pipelining that we already have?
I.e., just stick the `GeneralisedSignatureMethod` and classifier together.

**Comment by TonyBagnall** [comment]
I like to have specific classifiers for transformers for evaluation purposes. For example we put in the RocketClassifier so that we can concretely  point to the classifier we have evaluated. I would prefer it stay in. And @jambo6 well done on sticking with it, we really appreciate the effort involved

**Comment by jambo6** [comment]
SignatureClassifier is effectively redundant, but it was so as to keep inline with with other things all having Classifier at the end. I've no strong feelings with regards to how the signature module should be laid out though, so am happy to edit/remove anything if you ever want me to.

Do I need to do anything else for this at the moment?

**Comment by fkiraly** [comment]
This looks really great!

Looks almost ready to be merged.

Quick remarks:

* you should add new transformers and estimators to `sktime.tests._config` - add the new estimators, and some reasonable parameter combinations
* I like your decorators aroudn `fit` and `predict`. We've been internally disussing how to deal with the various conversions, see the related PR #980 and discussion for forecasting. Perhaps you have some good ideas?

File: .all-contributorsrc
Comment (mloning) [inline .all-contributorsrc:611]: Is this change intended?

File: sktime/classification/signature_based/_signature_classifier.py
Comment (mloning) [inline sktime/classification/signature_based/_signature_classifier.py:108]: I'd suggest to make this private
```suggestion
    def _setup_classification_pipeline(self):
```

File: sktime/transformations/panel/signature_based/_signature_method.py
Comment (mloning) [inline sktime/transformations/panel/signature_based/_signature_method.py:16]: Do we need this here?
```suggestion

```

File: sktime/transformations/panel/signature_based/_signature_method.py
Comment (mloning) [inline sktime/transformations/panel/signature_based/_signature_method.py:65]: Is this used?

**Review by mloning — COMMENTED** [review]
Hi @jambo6, really happy that you could sort out the PR! I just have a few minor comments.

Also please:
* add yourself to the `CODEOWNERS` file
* add the main classes to `sktime/docs/api_reference/`

**Comment by jambo6** [comment]
> * I like your decorators aroudn `fit` and `predict`. We've been internally disussing how to deal with the various conversions, see the related PR [Forecasting support for multivariate y and multiple input/output types - working prototype #980](https://github.com/alan-turing-institute/sktime/pull/980) and discussion for forecasting. Perhaps you have some good ideas?

Perhaps a file that has a selection of decorators that provide conversions for fit/transform methods to a specific type? For example the decorators convert any pandas input to numpy and convert back. It would be good to have some logic that can convert between sktime formats under the hood, so the code in any fit/transform only needs to consider one particular format. 

@mloning I have updated the docs and made the edits according to your comments.

**Review by mloning — APPROVED** [review]
Thanks for the changes @jambo6 - I'm happy to merge this now once the tests all pass! Does anyone else want to review before we merge?

**Comment by fkiraly** [comment]
but the tests don't all pass! It somehow destroys the test examples

**Comment by fkiraly** [comment]
it's just some import error that should be easily fixable, @jambo6

**Review by fkiraly — CHANGES_REQUESTED** [review]
this is massive but changes are localized and interface compliant.

Looks good, except for the import error mentioned above, which needs fixing before merge.

File: build_tools/run_examples.sh
Comment (mloning) [inline build_tools/run_examples.sh:8]: That change is probably not intended?
```suggestion
for notebook in examples/*.ipynb; do
```

**Review by mloning — CHANGES_REQUESTED** [review]

**Comment by TonyBagnall** [comment]
sorry, clustering has introduced minor conflicts I think

**Review by mloning — APPROVED** [review]
I've had a look twice in the last few days, it's a lot of code, but I'm happy to merge this now. 

Really appreciate your work here @jambo6, especially after all the issues with the dependencies!  

Let me know if anyone else wants to review before we merge?

**Review by TonyBagnall — APPROVED** [review]
great job!

**Review by fkiraly — APPROVED** [review]
yes, looks good and tests seem to run

**Comment by mloning** [comment]
Now merged @jambo6 - thanks again, great work!

**Pull request merged**
